kotori/daq/storage/cratedb.py

```diff
@@ -0,0 +1,270 @@
+# -*- coding: utf-8 -*-
+# (c) 2023 Andreas Motl <andreas@getkotori.org>
+import calendar
+import functools
+import json
+from collections import OrderedDict
+from decimal import Decimal
+from copy import deepcopy
+from datetime import datetime, date, timezone
+
+import crate.client.http
+import pytz
+import requests
+from crate import client
+from crate.client.converter import DefaultTypeConverter
+from crate.client.exceptions import ProgrammingError
+from funcy import project
+from munch import Munch
+from twisted.logger import Logger
+
+from kotori.daq.storage.util import format_chunk
+
+log = Logger()
+
+
+class CrateDBAdapter:
+    """
+    Kotori database backend adapter for CrateDB.
+
+    CrateDB is a distributed SQL database for storing and analyzing
+    massive amounts of data in real-time. Built on top of Lucene.
+
+    https://github.com/crate/crate
+    """
+
+    def __init__(self, settings=None, database=None):
+        """
+        Carry over connectivity parameters.
+
+        TODO: Verify with CrateDB Cloud.
+        """
+
+        settings = deepcopy(settings) or {}
+        settings.setdefault("host", "localhost")
+        settings.setdefault("port", "4200")
+        settings.setdefault("username", "crate")
+        settings.setdefault("password", "")
+        settings.setdefault("database", database)
+
+        # TODO: Bring back pool size configuration.
+        # settings.setdefault('pool_size', 10)
+
+        settings["port"] = int(settings["port"])
+
+        # FIXME: This is bad style. Well, but it is currently
+        #        inherited from ~10 year old code, so c'est la vie.
+        self.__dict__.update(**settings)
+
+        # Bookkeeping for all databases having been touched already
+        self.databases_written_once = set()
+
+        self.host_uri = "{host}:{port}".format(**self.__dict__)
+
+        # TODO: Bring back pool size configuration.
+        # log.info('Storage target is {uri}, pool size is {pool_size}', uri=self.host_uri, pool_size=self.pool_size)
+        log.info("Storage target is {uri}", uri=self.host_uri)
+        self.db_client = client.connect(
+            self.host_uri, username=self.username, password=self.password, pool_size=20,
+        )
+
+    def get_tablename(self, meta):
+        """
+        Get table name for SensorWAN channel.
+        """
+        return f"{meta.database}.{meta.measurement}"
+
+    def create_table(self, tablename):
+        """
+        Create database table for SensorWAN channel.
+        """
+        log.info(f"Creating table: {tablename}")
+        sql_ddl = f"""
+CREATE TABLE IF NOT EXISTS {tablename} (
+    time TIMESTAMP WITH TIME ZONE DEFAULT NOW() NOT NULL, 
+    tags OBJECT(DYNAMIC), 
+    fields OBJECT(DYNAMIC)
+);      
```

Comment:
Will it be sensible to use such a DDL from the very beginning, @hlcianfagna?

```suggestion
CREATE TABLE IF NOT EXISTS {tablename} (
    time TIMESTAMP WITH TIME ZONE DEFAULT NOW() NOT NULL, 
    tags OBJECT(DYNAMIC), 
    fields OBJECT(DYNAMIC),
    year TIMESTAMP GENERATED ALWAYS AS DATE_TRUNC('year', time)
) PARTITIONED BY (year);
```

Reply:
I see you went for this, I think it is a good idea, yes